### ISSUE TITLE
Light & Material functions not compatible with beginShape() endShape() elements

### ISSUE THREAD

**Issue by whoem** [processing/p5.js#1180]
Chrome: "glDrawArrays: attempt to render with no buffer attached to enabled attribute 2"
Firefox: "WebGL: drawArrays: no VBO bound to enabled vertex attrib index 2!"

Would be veeeeeery happy to see this fixed. Looking forward to a full feature p5.js!!!! :) *excited*
Cheers!

**Comment by indefinit** [comment]
Thanks for filing.  This is a known issue and will be fixed in the next major update (in fact it's already fixed in my fork).  The reason for it taking so long is that immediate mode (beginShape and endShape) is undergoing a larger architectural change in the webgl core.

**Comment by whoem** [comment]
Sounds gooood :D You know, when will the next major update happen?
Do you have a build of your fork for me?

**Comment by whoem** [comment]
Hmm. After building your fork I get another error:
"Uncaught TypeError: Cannot read property '0' of undefined" ... "p5.js:8999"

**Comment by indefinit** [comment]
Do you mind posting your code or a link to a gist? I'm targeting getting this done before the new year.

**Comment by whoem** [comment]
The problem is present in the definition of function "p5.Renderer3D.prototype.background". Here you first define the variable "_col", further the variables "_r", "_g", "_b", "_a". Their definition references to a "_col.rgb" array. But this does not exist. In fact, "_col" has a "maxes" object, which contains a "rgb" array. So I fixed the four lines like this: "var _r = (_col.rgb[0]) / 255;" to "var _r = (_col.maxes.rgb[0]) / 255;". Now your fork build works here. But on canvas my beginShape/endShape cluster is an all over see-through stroke based construction on white background (eventhough I declared "noStroke()", "background(0)" and "endShape(CLOSE)"). "ambientLight ()" and "ambientMaterial()" still end up causing the exact same error mentioned at the beginning of this thread.

Here is my sketch:
```
var rotX, rotY, scale, cutWidth, cutMargin;

function setup() {
  createCanvas(500, 500, WEBGL);

  rotX = 0;
  rotY = 0;
  scale = 200;
  cutWidth = .1;
  cutMargin = cutWidth/sin(atan(sqrt(2)/2));
}

function draw() {
  noStroke();
  //ambientLight (255, 0, 0);
  background(0);

  //translate (width/2, height/2);
  rotateX (rotX);
  rotateY (rotY);

  //ambientMaterial (250);
  mainBody();
}

function mouseDragged() {
  var rate = 0.01;
  rotX += (pmouseY-mouseY) * rate;
  rotY += (mouseX-pmouseX) * rate;
}

function mainBody() {
  beginShape();
  vertex (-1*scale, -1*scale, 1*scale);
  vertex ((1-cutMargin)*scale, -1*scale, 1*scale);
  vertex (1*scale, -1*scale, (1-cutMargin)*scale);
  vertex (1*scale, -1*scale, -1*scale);
  vertex ((-1+cutMargin)*scale, -1*scale, -1*scale);
  vertex (-1*scale, -1*scale, (-1+cutMargin)*scale);
  endShape (CLOSE);

  beginShape();
  vertex (-1*scale, -1*scale, 1*scale);
  vertex ((1-cutMargin)*scale, -1*scale, 1*scale);
  vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
  endShape (CLOSE);

  beginShape();
  vertex (-1*scale, -1*scale, 1*scale);
  vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
  vertex (-1*scale, -1*scale, (-1+cutMargin)*scale);
  endShape (CLOSE);

  beginShape();
  vertex (1*scale, -1*scale, -1*scale);
  vertex ((-1+cutMargin)*scale, -1*scale, -1*scale);
  vertex (1*scale, (1-cutMargin)*scale, -1*scale);
  endShape (CLOSE);

  beginShape();
  vertex (1*scale, -1*scale, -1*scale);
  vertex (1*scale, (1-cutMargin)*scale, -1*scale);
  vertex (1*scale, -1*scale, (1-cutMargin)*scale);
  endShape (CLOSE);

  beginShape();
  vertex ((1-cutMargin)*scale, -1*scale, 1*scale);
  vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
  vertex (1*scale, (1-cutMargin)*scale, -1*scale);
  vertex (1*scale, -1*scale, (1-cutMargin)*scale);
  endShape (CLOSE);

  beginShape();
  vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
  vertex (-1*scale, -1*scale, (-1+cutMargin)*scale);
  vertex ((-1+cutMargin)*scale, -1*scale, -1*scale);
  vertex (1*scale, (1-cutMargin)*scale, -1*scale);
  endShape (CLOSE);
}
```
And there is a screenshot of what the browser creates:
![shape](https://cloud.githubusercontent.com/assets/6699384/11917923/6b4ac1f2-a71b-11e5-9c1e-85e54293afbe.png)

**Comment by indefinit** [comment]
Ok I'll take a look further.  It's possible the issue has to do with an undefined attribute for lighting in the pixel shader.  Will report back shortly.

**Comment by indefinit** [comment]
Ok it looks like there's some weirdness going on with background and color
in general, perhaps from a previous upstream merge.  I'll have to dig
deeper but in case you are curious or want to take a stab, check out my
geometry-refactor branch here.  PR's welcome:
https://github.com/indefinit/p5.js/tree/geometry-refactor



On Mon, Dec 21, 2015 at 2:47 PM, Kevin Siwoff <kevin.siwoff@gmail.com>
wrote:

> Ok I'll take a look further.  It's possible the issue has to do with an
> undefined attribute for lighting in the pixel shader.  Will report back
> shortly.
>
> Sent from my mobile.
>
> On Dec 20, 2015, at 1:13 PM, whoem <notifications@github.com> wrote:
>
> The problem is present in the definition of function
> "p5.Renderer3D.prototype.background". Here you first define the variable
> "_col", further the variables "_r", "_g", "_b", "_a". Their definition
> references to a "_col.rgb" array. But this does not exist. In fact, "_col"
> has a "maxes" object, which contains a "rgb" array. So I fixed the four
> lines like this: "var _r = (_col.rgb[0]) / 255;" to "var _r =
> (_col.maxes.rgb[0]) / 255;". Now your fork build works here. But on canvas
> my beginShape/endShape cluster is an all over see-through stroke based
> construction on white background (eventhough I declared "noStroke()",
> "background(0)" and "endShape(CLOSE)"). "ambientLight ()" and
> "ambientMaterial()" still end up causing the exact same error mentioned at
> the beginning of this thread.
>
> Here is my sketch:
>
> var rotX, rotY, scale, cutWidth, cutMargin;
>
> function setup() {
>   createCanvas(500, 500, WEBGL);
>
>   rotX = 0;
>   rotY = 0;
>   scale = 200;
>   cutWidth = .1;
>   cutMargin = cutWidth/sin(atan(sqrt(2)/2));
> }
>
> function draw() {
>   noStroke();
>   //ambientLight (255, 0, 0);
>   background(0);
>
>   //translate (width/2, height/2);
>   rotateX (rotX);
>   rotateY (rotY);
>
>   //ambientMaterial (250);
>   mainBody();
> }
>
> function mouseDragged() {
>   var rate = 0.01;
>   rotX += (pmouseY-mouseY) * rate;
>   rotY += (mouseX-pmouseX) * rate;
> }
>
> function mainBody() {
>   beginShape();
>   vertex (-1*scale, -1*scale, 1*scale);
>   vertex ((1-cutMargin)*scale, -1*scale, 1*scale);
>   vertex (1*scale, -1*scale, (1-cutMargin)*scale);
>   vertex (1*scale, -1*scale, -1*scale);
>   vertex ((-1+cutMargin)*scale, -1*scale, -1*scale);
>   vertex (-1*scale, -1*scale, (-1+cutMargin)*scale);
>   endShape (CLOSE);
>
>   beginShape();
>   vertex (-1*scale, -1*scale, 1*scale);
>   vertex ((1-cutMargin)*scale, -1*scale, 1*scale);
>   vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
>   endShape (CLOSE);
>
>   beginShape();
>   vertex (-1*scale, -1*scale, 1*scale);
>   vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
>   vertex (-1*scale, -1*scale, (-1+cutMargin)*scale);
>   endShape (CLOSE);
>
>   beginShape();
>   vertex (1*scale, -1*scale, -1*scale);
>   vertex ((-1+cutMargin)*scale, -1*scale, -1*scale);
>   vertex (1*scale, (1-cutMargin)*scale, -1*scale);
>   endShape (CLOSE);
>
>   beginShape();
>   vertex (1*scale, -1*scale, -1*scale);
>   vertex (1*scale, (1-cutMargin)*scale, -1*scale);
>   vertex (1*scale, -1*scale, (1-cutMargin)*scale);
>   endShape (CLOSE);
>
>   beginShape();
>   vertex ((1-cutMargin)*scale, -1*scale, 1*scale);
>   vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
>   vertex (1*scale, (1-cutMargin)*scale, -1*scale);
>   vertex (1*scale, -1*scale, (1-cutMargin)*scale);
>   endShape (CLOSE);
>
>   beginShape();
>   vertex (-1*scale, (1-cutMargin)*scale, 1*scale);
>   vertex (-1*scale, -1*scale, (-1+cutMargin)*scale);
>   vertex ((-1+cutMargin)*scale, -1*scale, -1*scale);
>   vertex (1*scale, (1-cutMargin)*scale, -1*scale);
>   endShape (CLOSE);
> }
>
> And there is a screenshot of what the browser creates:
> [image: shape]
> <https://cloud.githubusercontent.com/assets/6699384/11917923/6b4ac1f2-a71b-11e5-9c1e-85e54293afbe.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/processing/p5.js/issues/1180#issuecomment-166112690>.
>
>


-- 

Kevin Siwoff
// http://kevinsiwoff.com // Instagram: @kevinsiwoff
<http://instagram.com/kevinsiwoff/> //

**Comment by whoem** [comment]
Hey, sorry for late reply. I'd like to collaborate on processing.js - unfortunately I am very busy until end of march. Please let me know in this thread, if there are major changes that could fix the problems with my code :) Thank you

**Comment by indefinit** [comment]
This should be fixed with #1196 in the geometry-refactor branch.  Waiting on a couple of pending changes before submitting another PR and merging into the master.
